File: start/tokenizer.js

class Tokenizer {
  constructor(text) {
    this.text = text;
    this.pos = 0;
  }

  current() {
    if (this.pos >= this.text.length) {
      return null;
    }
    return this.text.charAt(this.pos);
  }

  skipSpaces() {
    while (this.current() === ' ' || this.current() === '\n') {
      this.pos++;
    }
  }

  run() {
    const tokens = [];

    for (;;) {
      this.skipSpaces();

      const c = this.current();
      if (c === null) {
        break;
      }

      throw `unrecognized: ${c}`;
    }
    return tokens;
  }
}


function test() {
  // {type: 'NUMBER', value: 2}, {type: '*'}, {type: 'NUMBER', value: 2}
  console.log(new Tokenizer('2 * 2').run());

  console.log(new Tokenizer('2   *   2').run());
  console.log(new Tokenizer('2\n+\n2').run());
  console.log(new Tokenizer('#comment\n 2 + 2').run());
}

test();


module.exports = { Tokenizer };

File: start/tokenizer.py
class Tokenizer:
    def __init__(self, text):
        self.text = text
        self.pos = 0

        self.tokens = []

    def current(self):
        if self.pos >= len(self.text):
            return None
        return self.text[self.pos]

    def skip_spaces(self):
        while self.current() == ' ' and self.current() == '\n':
            self.pos += 1

    def run(self):
        tokens = []

        while True:
            self.skip_spaces()

            c = self.current()
            # ...

            raise Exception(f'unexpected: {c}')

        return tokens


def test():
    print(Tokenizer('2 + (2 * 2)').run())


if __name__ == '__main__':
    test()
